### PR TITLE
Refactor read_PE

### PR DESCRIPTION
**Pull request by marcelm**
- Factor out a `rescue_read` function. This is called once for R1 and once for R2. The function has quite a long signature, but at least the code duplication is avoided. I have some ideas for how to reduce the number of parameters.
- Add extra reads to the test dataset that force the rescue code to be run (for both R1 and R2). Good that I did that, otherwise I wouldn’t have caught that I had swapped two of the parameters when calling `rescue_read`. 
- I still have to figure out what the difference is between `rescue_mate` and this new `rescue_read`, but that can be done later.
- Change the order of cases in `align_PE` in order to allow early returns. The cases where no reads map or only one of them (triggering rescue) are now "special cases" dealt with at the beginning of the function, and the "normal case" (NAMs for both R1 and R2 found) follows after, and can be de-dented.
- Somewhat unrelated, I factored out an `i_dist_est::update` function that updates the current insert size distribution estimate.

### CODE REVIEW

**Comment by marcelm** [comment]
Sorry for the big diff. This is mainly from de-denting lots of code in the function. You can make it more readable by clicking on the cog wheel in the "Files changed" view and enabling "Hide whitespace".

**Comment by ksahlin** [comment]
From memory; 

`Rescue_read:` This function is called for reads that had most seeds hard-masked (repetitive >1000 times in index i think) This function call is performed at the seeding step, and can be called both in SE and PE mode. For very abundant seeds, the rescure_read function is optimised in processing for higher abundance seeds.

`Rescure mate:` This function is called in the extension step only for PE reads, when one read mapped but the mate lacks seeds. Strobealign is basically guessing that the mate could be in some window upstream or downstream from the read (based on orientation) and does an alignment to the reference in a larger segment (fairly costly).

**Comment by marcelm** [comment]
The `rescue_read` function is the one I factored out in this PR, so now that I’ve figured out which functions call which (#98), I think you mean the `find_nams_rescue` function instead. It fits your description in that it is called in both SE and PE mode. I’ll merge this for now and then we can come up with a better name later.

**Comment by ksahlin** [comment]
You are right, I confused `Rescue_read` with `find_nams_rescue`. Also, about my comment on the if-statements being flipped, I think it is just the comment below the if-statement that is wrong (flipped). The code looks right.

**Comment by marcelm** [comment]
Sorry, I did not see your comment on the if statements. In case you commented on the code and clicked on 'Start a review', note that the comments are only submitted once you also click on 'Submit review".

File: src/aln.cpp [hunk 1]
```diff
@@ -2344,216 +2344,264 @@ inline void align_PE(alignment_params &aln_params, Sam &sam, std::vector<nam> &a
...
     bool rc_already_comp2 = false;
+    double secondary_dropoff = 2 * aln_params.mismatch + aln_params.gap_open;
+
+    if (all_nams1.empty() && all_nams2.empty()) {
+         // None of the reads have any NAMs
+        sam.add_unmapped_pair(record1, record2);
+        return;
+    }
+
+    if (!all_nams1.empty() && all_nams2.empty()) {
```
[the remark this answers is not in the document]
Comment (marcelm) [reply src/aln.cpp:2359]: I’ll fix this

**Pull request merged**